flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/ZooKeeperSubmittedJobGraphStore.java

```diff
@@ -0,0 +1,380 @@
+		pathCache.getListenable().addListener(new SubmittedJobGraphsPathCacheListener());
+	}
+
+	@Override
+	public void start(SubmittedJobGraphListener jobGraphListener) throws Exception {
+		synchronized (cacheLock) {
+			if (!isRunning) {
+				this.jobGraphListener = jobGraphListener;
+
+				pathCache.start();
+
+				isRunning = true;
+			}
+		}
+	}
+
+	@Override
+	public void stop() throws Exception {
+		synchronized (cacheLock) {
+			if (isRunning) {
+				jobGraphListener = null;
+
+				pathCache.close();
+
+				client.close();
+
+				isRunning = false;
+			}
+		}
+	}
+
+	@Override
+	public List<SubmittedJobGraph> recoverJobGraphs() throws Exception {
+		synchronized (cacheLock) {
+			verifyIsRunning();
+
+			List<ZooKeeperStateHandleStore.ZooKeeperStateHandle<SubmittedJobGraph>> submitted;
+
+			while (true) {
+				try {
+					submitted = jobGraphsInZooKeeper.getAll();
+					break;
+				}
+				catch (ConcurrentModificationException e) {
+					LOG.warn("Concurrent modification while reading from ZooKeeper. Retrying.");
+				}
+			}
+
+			if (submitted.size() != 0) {
+				List<SubmittedJobGraph> jobGraphs = new ArrayList<>(submitted.size());
+
+				for (ZooKeeperStateHandleStore.ZooKeeperStateHandle<SubmittedJobGraph> jobStateHandle : submitted) {
+					SubmittedJobGraph jobGraph = jobStateHandle
+							.getState(ClassLoader.getSystemClassLoader());
+
+					addedJobGraphs.add(jobGraph.getJobId());
+
+					jobGraphs.add(jobGraph);
+				}
+
+				LOG.info("Recovered {} job graphs: {}.", jobGraphs.size(), jobGraphs);
+				return jobGraphs;
+			}
+			else {
+				LOG.info("No job graph to recover.");
+				return Collections.emptyList();
+			}
+		}
+	}
+
+	@Override
+	public Option<SubmittedJobGraph> recoverJobGraph(JobID jobId) throws Exception {
+		checkNotNull(jobId, "Job ID");
+		String path = getPathForJob(jobId);
+
+		synchronized (cacheLock) {
+			verifyIsRunning();
+
+			try {
+				StateHandle<SubmittedJobGraph> jobStateHandle = jobGraphsInZooKeeper.get(path);
+
+				SubmittedJobGraph jobGraph = jobStateHandle
+						.getState(ClassLoader.getSystemClassLoader());
+
+				addedJobGraphs.add(jobGraph.getJobId());
+
+				LOG.info("Recovered {}.", jobGraph);
+
+				return Option.apply(jobGraph);
+			}
+			catch (KeeperException.NoNodeException ignored) {
+			}
+
+			return Option.empty();
+		}
+	}
+
+	@Override
+	public void putJobGraph(SubmittedJobGraph jobGraph) throws Exception {
+		checkNotNull(jobGraph, "Job graph");
+		String path = getPathForJob(jobGraph.getJobId());
+
+		boolean success = false;
+
+		while (!success) {
+			synchronized (cacheLock) {
+				verifyIsRunning();
+
+				int currentVersion = jobGraphsInZooKeeper.exists(path);
+
+				if (currentVersion == -1) {
+					try {
+						jobGraphsInZooKeeper.add(path, jobGraph);
+						LOG.info("Added {} to ZooKeeper.", jobGraph);
+
+						success = true;
+					}
+					catch (KeeperException.NodeExistsException ignored) {
+						continue;
```

Comment:
No need for `continue`. You can simply check `if (success)` before the `addedJobGraphs.add(jobGraph.getJobId())`.